File: zig/src/stream.zig
//! Stream - A multiplexed reliable stream over KCP.

const std = @import("std");
const kcp = @import("kcp.zig");
const ring_buffer = @import("ring_buffer.zig");

/// Re-export RingBuffer for convenience
pub const RingBuffer = ring_buffer.RingBuffer;

/// Stream state
pub const StreamState = enum {
    open,
    local_close, // We sent FIN
    remote_close, // We received FIN
    closed,
};

/// Stream errors
pub const StreamError = error{
    StreamClosed,
    KcpSendFailed,
    Timeout,
};

/// Stream represents a multiplexed stream over KCP.
pub const Stream = struct {
    id: u32,
    mux: *Mux,
    kcp_instance: ?kcp.Kcp,
    state: StreamState,
    recv_buf: RingBuffer(u8), // O(1) head removal
    allocator: std.mem.Allocator,
    output_error: bool, // Set on transport output error

    /// Initialize a new stream.
    pub fn init(allocator: std.mem.Allocator, id: u32, mux: *Mux) !*Stream {
        const self = try allocator.create(Stream);
        errdefer allocator.destroy(self);

        self.* = Stream{
            .id = id,
            .mux = mux,
            .kcp_instance = null,
            .state = .open,
            .recv_buf = RingBuffer(u8).init(allocator),
            .allocator = allocator,
            .output_error = false,
        };

        // Create KCP with output callback that routes data through the Mux
        self.kcp_instance = try kcp.Kcp.init(id, &Stream.kcpOutput, self);
        if (self.kcp_instance) |*k| {
            k.setUserPtr(); // Set user pointer for callback to access this Stream
            k.setDefaultConfig();
        }

        return self;
    }

    /// Deinitialize the stream.
    pub fn deinit(self: *Stream) void {
        if (self.kcp_instance) |*k| {
            k.deinit();
        }
        self.recv_buf.deinit();
        self.allocator.destroy(self);
    }

    /// Get stream ID.
    pub fn getId(self: *const Stream) u32 {
        return self.id;
    }

    /// Get stream state.
    pub fn getState(self: *const Stream) StreamState {
        return self.state;
    }

    /// Write data to the stream.
    pub fn write(self: *Stream, data: []const u8) StreamError!usize {
        if (self.state == .closed or self.state == .local_close) {
            return StreamError.StreamClosed;
        }

        if (self.kcp_instance) |*k| {
            const n = k.send(data);
            if (n < 0) {
                return StreamError.KcpSendFailed;
            }
            return @intCast(n);
        }

        return StreamError.KcpSendFailed;
    }

    /// Read data from the stream.
    pub fn read(self: *Stream, buffer: []u8) StreamError!usize {
        if (self.recv_buf.readableLength() == 0) {
            if (self.state == .closed or self.state == .remote_close) {
                return 0; // EOF
            }
            return 0; // No data available
        }

        // LinearFifo.read() automatically removes data from head - O(1)
        return self.recv_buf.read(buffer);
    }

    /// Receive data directly from KCP into user-provided buffer (zero-copy fast path).
    /// Returns the number of bytes received, or 0 if no data available.
    /// This bypasses the internal recv_buf for lower latency when caller can process immediately.
    pub fn recvIntoBuffer(self: *Stream, buffer: []u8) StreamError!usize {
        // First drain any buffered data
        if (self.recv_buf.readableLength() > 0) {
            return self.recv_buf.read(buffer);
        }

        // No buffered data, try direct receive from KCP
        if (self.kcp_instance) |*k| {
            const size = k.peekSize();
            if (size <= 0) {
                if (self.state == .closed or self.state == .remote_close) {
                    return 0; // EOF
                }
                return 0; // No data available
            }

            // Direct receive into user buffer if it fits
            if (buffer.len >= @as(usize, @intCast(size))) {
                const n = k.recv(buffer);
                if (n > 0) {
                    return @intCast(n);
                }
            } else {
                // Buffer too small, need intermediate allocation
                const tmp = try self.allocator.alloc(u8, @intCast(size));
                defer self.allocator.free(tmp);

                const n = k.recv(tmp);
                if (n > 0) {
                    const copy_len = @min(buffer.len, @as(usize, @intCast(n)));
                    @memcpy(buffer[0..copy_len], tmp[0..copy_len]);
                    // Buffer remaining data
                    if (@as(usize, @intCast(n)) > copy_len) {
                        try self.recv_buf.write(tmp[copy_len..@intCast(n)]);
                    }
                    return copy_len;
                }
            }
        }

        return 0;
    }

    /// Close the stream.
    /// Shutdown the write-half of the stream.
    /// Sends a FIN frame to the remote peer and transitions to `local_close` state.
    /// The stream can still receive data until a FIN is received from the peer.
    pub fn shutdown(self: *Stream) void {
        if (self.state == .closed or self.state == .local_close) return;

        // Only send FIN when transitioning from open to local_close
        const should_send_fin = self.state == .open;

        if (self.state == .open) {
            self.state = .local_close;
        } else {
            self.state = .closed;
        }

        // Send FIN (only once)
        if (should_send_fin) {
            self.mux.sendFin(self.id) catch {};
        }
    }

    /// Check if a transport output error has occurred.
    pub fn hasOutputError(self: *const Stream) bool {
        return self.output_error;
    }

    /// Input data from KCP.
    pub fn kcpInput(self: *Stream, data: []const u8) void {
        if (self.state == .closed) return;
        if (self.kcp_instance) |*k| {
            _ = k.input(data);
        }
    }

    /// Receive data from KCP and buffer it.
    /// Uses stack buffer for common MTU-sized messages, heap fallback for oversized.
    /// Heap buffer is reused across loop iterations to avoid repeated allocations.
    /// Returns true if any data was received.
    pub fn kcpRecv(self: *Stream) bool {
        var received = false;
        if (self.kcp_instance) |*k| {
            // Reusable MTU-sized stack buffer for common case
            var stack_buf: [1500]u8 = undefined;
            // Reusable heap buffer for oversized messages
            var heap_buf: ?[]u8 = null;
            defer if (heap_buf) |buf| self.allocator.free(buf);

            while (true) {
                const size = k.peekSize();
                if (size <= 0) break;

                const usize_size: usize = @intCast(size);

                if (usize_size <= stack_buf.len) {
                    // Common path: use stack buffer (no allocation)
                    const n = k.recv(stack_buf[0..usize_size]);
                    if (n <= 0) break;
                    self.recv_buf.write(stack_buf[0..@intCast(n)]) catch break;
                    received = true;
                } else {
                    // Rare path: reuse or grow heap buffer for oversized messages
                    if (heap_buf == null or heap_buf.?.len < usize_size) {
                        if (heap_buf) |old| self.allocator.free(old);
                        heap_buf = self.allocator.alloc(u8, usize_size) catch break;
                    }
                    const n = k.recv(heap_buf.?[0..usize_size]);
                    if (n <= 0) break;
                    self.recv_buf.write(heap_buf.?[0..@intCast(n)]) catch break;
                    received = true;
                }
            }
        }
        return received;
    }

    /// Update KCP state.
    pub fn kcpUpdate(self: *Stream, current: u32) void {
        if (self.state == .closed) return;
        if (self.kcp_instance) |*k| {
            k.update(current);
        }
    }

    /// Handle FIN from remote.
    pub fn handleFin(self: *Stream) void {
        if (self.state == .local_close) {
            self.state = .closed;
        } else if (self.state == .open) {
            self.state = .remote_close;
        }
    }

    /// KCP output callback
    fn kcpOutput(data: []const u8, user: ?*anyopaque) void {
        if (user) |u| {
            const stream: *Stream = @ptrCast(@alignCast(u));
            stream.mux.sendPsh(stream.id, data) catch |err| {
                stream.output_error = true;
                std.log.err("Mux output error in stream {d}: {s}", .{ stream.id, @errorName(err) });
            };
        }
    }
};

/// Mux configuration
pub const MuxConfig = struct {
    max_frame_size: usize = 32 * 1024,
    max_receive_buffer: usize = 256 * 1024,
};

/// Callback when stream has data available to read.
pub const OnStreamDataFn = *const fn (stream_id: u32) void;

/// Callback when a new stream is accepted.
pub const OnNewStreamFn = *const fn (stream: *Stream) void;

/// Mux multiplexes multiple streams over a single connection.
pub const Mux = struct {
    config: MuxConfig,
    output_fn: *const fn ([]const u8) anyerror!void,
    on_stream_data: OnStreamDataFn,
    on_new_stream: OnNewStreamFn,
    is_client: bool,
    streams: std.AutoHashMap(u32, *Stream),
    next_id: u32,
    closed: bool,
    allocator: std.mem.Allocator,

    /// Initialize a new Mux.
    pub fn init(
        allocator: std.mem.Allocator,
        config: MuxConfig,
        is_client: bool,
        output_fn: *const fn ([]const u8) anyerror!void,
        on_stream_data: OnStreamDataFn,
        on_new_stream: OnNewStreamFn,
    ) !*Mux {
        const self = try allocator.create(Mux);
        errdefer allocator.destroy(self);

        self.* = Mux{
            .config = config,
            .output_fn = output_fn,
            .on_stream_data = on_stream_data,
            .on_new_stream = on_new_stream,
            .is_client = is_client,
            .streams = .init(allocator),
            .next_id = if (is_client) 1 else 2, // Client: odd, Server: even
            .closed = false,
            .allocator = allocator,
        };

        return self;
    }

    /// Deinitialize the Mux.
    pub fn deinit(self: *Mux) void {
        // Close all streams
        var iter = self.streams.valueIterator();
        while (iter.next()) |stream| {
            stream.*.deinit();
        }
        self.streams.deinit();
        self.allocator.destroy(self);
    }

    /// Open a new stream.
    pub fn openStream(self: *Mux) !*Stream {
        if (self.closed) return error.MuxClosed;

        const id = self.next_id;
        self.next_id += 2;

        const stream = try Stream.init(self.allocator, id, self);
        errdefer stream.deinit();

        try self.streams.put(id, stream);
        errdefer _ = self.streams.remove(id); // Remove from map before deinit to avoid dangling pointer

        // Send SYN
        try self.sendSyn(id);

        return stream;
    }

    /// Get number of active streams.
    pub fn numStreams(self: *const Mux) usize {
        return self.streams.count();
    }

    /// Check if closed.
    pub fn isClosed(self: *const Mux) bool {
        return self.closed;
    }

    /// Close the Mux.
    pub fn closeMux(self: *Mux) void {
        self.closed = true;
    }

    /// Input a frame.
    pub fn input(self: *Mux, data: []const u8) !void {
        if (self.closed) return error.MuxClosed;

        const frame = try kcp.Frame.decode(data);

        switch (frame.cmd) {
            .syn => try self.handleSyn(frame.stream_id),
            .fin => self.handleFin(frame.stream_id),
            .psh => self.handlePsh(frame.stream_id, frame.payload),
            .nop => {}, // Keepalive, nothing to do
        }
    }

    /// Handle SYN (stream open request).
    fn handleSyn(self: *Mux, id: u32) !void {
        if (self.streams.contains(id)) return; // Duplicate SYN

        const stream = try Stream.init(self.allocator, id, self);
        errdefer stream.deinit();

        try self.streams.put(id, stream);

        // Notify via callback
        self.on_new_stream(stream);
    }

    /// Handle FIN (stream close).
    fn handleFin(self: *Mux, id: u32) void {
        if (self.streams.get(id)) |stream| {
            stream.handleFin();
        }
    }

    /// Handle PSH (data).
    fn handlePsh(self: *Mux, id: u32, payload: []const u8) void {
        if (self.streams.get(id)) |stream| {
            stream.kcpInput(payload);
            if (stream.kcpRecv()) {
                self.on_stream_data(id);
            }
        }
    }

    /// Send a SYN frame.
    fn sendSyn(self: *Mux, id: u32) !void {
        try self.sendFrame(.{
            .cmd = .syn,
            .stream_id = id,
            .payload = &[_]u8{},
        });
    }

    /// Send a FIN frame.
    pub fn sendFin(self: *Mux, id: u32) !void {
        try self.sendFrame(.{
            .cmd = .fin,
            .stream_id = id,
            .payload = &[_]u8{},
        });
    }

    /// Send a PSH frame.
    pub fn sendPsh(self: *Mux, id: u32, payload: []const u8) !void {
        try self.sendFrame(.{
            .cmd = .psh,
            .stream_id = id,
            .payload = payload,
        });
    }

    /// Send a frame.
    fn sendFrame(self: *Mux, frame: kcp.Frame) !void {
        if (self.closed) return error.MuxClosed;

        const required_size = kcp.FrameHeaderSize + frame.payload.len;
        var stack_buf: [1500]u8 = undefined; // MTU-sized stack buffer

        if (required_size <= stack_buf.len) {
            // Use stack buffer for typical small frames
            const encoded = try frame.encode(&stack_buf);
            try self.output_fn(encoded);
        } else {
            // Heap allocate for oversized frames
            const buf = try self.allocator.alloc(u8, required_size);
            defer self.allocator.free(buf);
            const encoded = try frame.encode(buf);
            try self.output_fn(encoded);
        }
    }

    /// Remove a stream from the Mux.
    pub fn removeStream(self: *Mux, id: u32) void {
        _ = self.streams.remove(id);
    }

    /// Update all streams.
    pub fn update(self: *Mux, current: u32) void {
        var iter = self.streams.iterator();
        while (iter.next()) |entry| {
            const id = entry.key_ptr.*;
            const stream = entry.value_ptr.*;
            stream.kcpUpdate(current);
            if (stream.kcpRecv()) {
                self.on_stream_data(id);
            }
        }
    }
};

// Tests
test "Frame encode decode" {
    const frame = kcp.Frame{
        .cmd = .syn,
        .stream_id = 1,
        .payload = &[_]u8{},
    };
    var buf: [kcp.FrameHeaderSize]u8 = undefined;
    const encoded = try frame.encode(&buf);
    const decoded = try kcp.Frame.decode(encoded);
    try std.testing.expectEqual(kcp.Cmd.syn, decoded.cmd);
    try std.testing.expectEqual(@as(u32, 1), decoded.stream_id);
}

test "Mux init deinit" {
    const allocator = std.testing.allocator;

    const outputFn = struct {
        fn output(_: []const u8) anyerror!void {}
    }.output;
    const onStreamData = struct {
        fn cb(_: u32) void {}
    }.cb;
    const onNewStream = struct {
        fn cb(_: *Stream) void {}
    }.cb;

    const mux = try Mux.init(allocator, .{}, true, outputFn, onStreamData, onNewStream);
    mux.deinit();
}

test "Mux open stream" {
    const allocator = std.testing.allocator;

    const outputFn = struct {
        fn output(_: []const u8) anyerror!void {}
    }.output;
    const onStreamData = struct {
        fn cb(_: u32) void {}
    }.cb;
    const onNewStream = struct {
        fn cb(_: *Stream) void {}
    }.cb;

    const mux = try Mux.init(allocator, .{}, true, outputFn, onStreamData, onNewStream);
    defer mux.deinit();

    const stream = try mux.openStream();
    try std.testing.expectEqual(@as(u32, 1), stream.getId());
    try std.testing.expectEqual(@as(usize, 1), mux.numStreams());
}

test "Stream state" {
    const allocator = std.testing.allocator;

    const outputFn = struct {
        fn output(_: []const u8) anyerror!void {}
    }.output;
    const onStreamData = struct {
        fn cb(_: u32) void {}
    }.cb;
    const onNewStream = struct {
        fn cb(_: *Stream) void {}
    }.cb;

    const mux = try Mux.init(allocator, .{}, true, outputFn, onStreamData, onNewStream);
    defer mux.deinit();

    const stream = try mux.openStream();
    try std.testing.expectEqual(StreamState.open, stream.getState());
}

File: rust/src/stream.rs
//! Stream - A multiplexed reliable stream over KCP.

use std::collections::{HashMap, VecDeque};
use std::io;
use std::sync::{Arc, Mutex, RwLock};

use crate::kcp::{Cmd, Frame, Kcp};

/// Stream state
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StreamState {
    Init,
    Open,
    LocalClose,  // We sent FIN
    RemoteClose, // We received FIN
    Closed,
}

/// Stream errors
#[derive(Debug)]
pub enum StreamError {
    StreamClosed,
    KcpSendFailed,
    Timeout,
    IoError(io::Error),
}

impl std::fmt::Display for StreamError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            StreamError::StreamClosed => write!(f, "stream closed"),
            StreamError::KcpSendFailed => write!(f, "KCP send failed"),
            StreamError::Timeout => write!(f, "timeout"),
            StreamError::IoError(e) => write!(f, "IO error: {}", e),
        }
    }
}

impl std::error::Error for StreamError {}

impl From<io::Error> for StreamError {
    fn from(e: io::Error) -> Self {
        StreamError::IoError(e)
    }
}

/// Stream represents a multiplexed stream over KCP.
pub struct Stream {
    id: u32,
    kcp: Mutex<Kcp>,
    state: RwLock<StreamState>,
    recv_buf: Mutex<VecDeque<u8>>, // O(1) head removal
}

impl Stream {
    /// Create a new stream.
    #[allow(clippy::type_complexity)]
    pub(crate) fn new(id: u32, output: Box<dyn Fn(&[u8]) + Send + Sync>) -> Self {
        let mut kcp = Kcp::new(id, output);
        kcp.set_default_config();

        Stream {
            id,
            kcp: Mutex::new(kcp),
            state: RwLock::new(StreamState::Open),
            recv_buf: Mutex::new(VecDeque::new()),
        }
    }

    /// Get stream ID.
    pub fn id(&self) -> u32 {
        self.id
    }

    /// Get stream state.
    pub fn state(&self) -> StreamState {
        *self.state.read().unwrap()
    }

    /// Write data to the stream.
    pub fn write_data(&self, data: &[u8]) -> Result<usize, StreamError> {
        let state = self.state();
        if state == StreamState::Closed || state == StreamState::LocalClose {
            return Err(StreamError::StreamClosed);
        }

        let mut kcp = self.kcp.lock().unwrap();
        let n = kcp.send(data);
        if n < 0 {
            return Err(StreamError::KcpSendFailed);
        }

        Ok(n as usize)
    }

    /// Read data from the stream.
    pub fn read_data(&self, buf: &mut [u8]) -> Result<usize, StreamError> {
        let mut recv_buf = self.recv_buf.lock().unwrap();

        if recv_buf.is_empty() {
            let state = self.state();
            if state == StreamState::Closed || state == StreamState::RemoteClose {
                return Ok(0); // EOF
            }
            return Ok(0); // No data available
        }

        // Use as_slices() for efficient bulk copy instead of byte-by-byte
        let (s1, s2) = recv_buf.as_slices();
        let n1 = std::cmp::min(buf.len(), s1.len());
        buf[..n1].copy_from_slice(&s1[..n1]);

        let n2 = std::cmp::min(buf.len().saturating_sub(n1), s2.len());
        if n2 > 0 {
            buf[n1..n1 + n2].copy_from_slice(&s2[..n2]);
        }

        let total_read = n1 + n2;
        recv_buf.drain(..total_read);

        Ok(total_read)
    }

    /// Close the stream.
    pub fn close(&self) {
        let mut state = self.state.write().unwrap();
        if *state == StreamState::Closed {
            return;
        }

        if *state == StreamState::Open {
            *state = StreamState::LocalClose;
        } else {
            *state = StreamState::Closed;
        }
    }

    /// Input data from KCP.
    pub(crate) fn kcp_input(&self, data: &[u8]) {
        if self.state() == StreamState::Closed {
            return;
        }

        let mut kcp = self.kcp.lock().unwrap();
        kcp.input(data);
    }

    /// Receive data from KCP and buffer it.
    pub(crate) fn kcp_recv(&self) {
        let mut kcp = self.kcp.lock().unwrap();
        let mut recv_buf = self.recv_buf.lock().unwrap();

        loop {
            let size = kcp.peek_size();
            if size <= 0 {
                break;
            }

            // Allocate buffer on heap based on actual message size
            let mut buf = vec![0u8; size as usize];
            let n = kcp.recv(&mut buf);
            if n <= 0 {
                break;
            }

            recv_buf.extend(&buf[..n as usize]);
        }
    }

    /// Receive data directly from KCP into user-provided buffer (zero-copy fast path).
    /// Returns the number of bytes received, or 0 if no data available.
    /// This bypasses the internal recv_buf for lower latency when caller can process immediately.
    pub fn recv_into_buffer(&self, buf: &mut [u8]) -> Result<usize, StreamError> {
        // First drain any buffered data
        {
            let mut recv_buf = self.recv_buf.lock().unwrap();
            if !recv_buf.is_empty() {
                let (s1, s2) = recv_buf.as_slices();
                let n1 = std::cmp::min(buf.len(), s1.len());
                buf[..n1].copy_from_slice(&s1[..n1]);

                let n2 = std::cmp::min(buf.len().saturating_sub(n1), s2.len());
                if n2 > 0 {
                    buf[n1..n1 + n2].copy_from_slice(&s2[..n2]);
                }

                let total_read = n1 + n2;
                recv_buf.drain(..total_read);
                return Ok(total_read);
            }
        }

        // No buffered data, try direct receive from KCP
        let mut kcp = self.kcp.lock().unwrap();
        let size = kcp.peek_size();
        if size <= 0 {
            let state = self.state();
            if state == StreamState::Closed || state == StreamState::RemoteClose {
                return Ok(0); // EOF
            }
            return Ok(0); // No data available
        }

        // Direct receive into user buffer if it fits
        if buf.len() >= size as usize {
            let n = kcp.recv(buf);
            if n > 0 {
                return Ok(n as usize);
            }
        } else {
            // Buffer too small, need intermediate allocation
            let mut tmp = vec![0u8; size as usize];
            let n = kcp.recv(&mut tmp);
            if n > 0 {
                let copy_len = std::cmp::min(buf.len(), n as usize);
                buf[..copy_len].copy_from_slice(&tmp[..copy_len]);
                // Buffer remaining data
                if n as usize > copy_len {
                    let mut recv_buf = self.recv_buf.lock().unwrap();
                    recv_buf.extend(&tmp[copy_len..n as usize]);
                }
                return Ok(copy_len);
            }
        }

        Ok(0)
    }

    /// Update KCP state.
    pub(crate) fn kcp_update(&self, current: u32) {
        if self.state() == StreamState::Closed {
            return;
        }

        let mut kcp = self.kcp.lock().unwrap();
        kcp.update(current);
    }

    /// Handle FIN from remote.
    pub(crate) fn handle_fin(&self) {
        let mut state = self.state.write().unwrap();
        if *state == StreamState::LocalClose {
            *state = StreamState::Closed;
        } else if *state == StreamState::Open {
            *state = StreamState::RemoteClose;
        }
    }
}

/// Mux configuration
#[derive(Debug, Clone)]
pub struct MuxConfig {
    pub max_frame_size: usize,
    pub max_receive_buffer: usize,
    pub accept_backlog: usize,
}

impl Default for MuxConfig {
    fn default() -> Self {
        MuxConfig {
            max_frame_size: 32 * 1024,
            max_receive_buffer: 256 * 1024,
            accept_backlog: 256,
        }
    }
}

/// Output function type for Mux
pub type OutputFn = Box<dyn Fn(&[u8]) -> Result<(), Box<dyn std::error::Error + Send + Sync>> + Send + Sync>;

/// Mux multiplexes multiple streams over a single connection.
pub struct Mux {
    #[allow(dead_code)]
    config: MuxConfig,
    output: Arc<OutputFn>,
    #[allow(dead_code)]
    is_client: bool,
    streams: RwLock<HashMap<u32, Arc<Stream>>>,
    next_id: Mutex<u32>,
    accept_queue: Mutex<Vec<Arc<Stream>>>,
    closed: RwLock<bool>,
}

impl Mux {
    /// Create a new Mux.
    pub fn new(config: MuxConfig, is_client: bool, output: OutputFn) -> Self {
        Mux {
            config,
            output: Arc::new(output),
            is_client,
            streams: RwLock::new(HashMap::new()),
            next_id: Mutex::new(if is_client { 1 } else { 2 }),
            accept_queue: Mutex::new(Vec::new()),
            closed: RwLock::new(false),
        }
    }

    /// Open a new stream.
    pub fn open_stream(&self) -> Result<Arc<Stream>, MuxError> {
        if self.is_closed() {
            return Err(MuxError::MuxClosed);
        }

        let id = {
            let mut next_id = self.next_id.lock().unwrap();
            let id = *next_id;
            *next_id += 2;
            id
        };

        let output = self.output.clone();
        let stream = Arc::new(Stream::new(
            id,
            Box::new(move |data| {
                // Use encode_with_payload to avoid intermediate allocations
                let encoded = Frame::encode_with_payload(Cmd::Psh, id, data);
                if let Err(e) = output(&encoded) {
                    eprintln!("Mux output error: {}", e);
                }
            }),
        ));

        self.streams.write().unwrap().insert(id, stream.clone());

        // Send SYN
        self.send_syn(id)?;

        Ok(stream)
    }

    /// Accept an incoming stream.
    pub fn accept_stream(&self) -> Result<Arc<Stream>, MuxError> {
        if self.is_closed() {
            return Err(MuxError::MuxClosed);
        }

        let mut queue = self.accept_queue.lock().unwrap();
        if queue.is_empty() {
            return Err(MuxError::NoStreamAvailable);
        }

        Ok(queue.remove(0))
    }

    /// Get number of active streams.
    pub fn num_streams(&self) -> usize {
        self.streams.read().unwrap().len()
    }

    /// Check if closed.
    pub fn is_closed(&self) -> bool {
        *self.closed.read().unwrap()
    }

    /// Close the Mux.
    pub fn close(&self) {
        *self.closed.write().unwrap() = true;
    }

    /// Input a frame.
    pub fn input(&self, data: &[u8]) -> Result<(), MuxError> {
        if self.is_closed() {
            return Err(MuxError::MuxClosed);
        }

        let frame = Frame::decode(data).map_err(|_| MuxError::InvalidFrame)?;

        match frame.cmd {
            Cmd::Syn => self.handle_syn(frame.stream_id)?,
            Cmd::Fin => self.handle_fin(frame.stream_id),
            Cmd::Psh => self.handle_psh(frame.stream_id, &frame.payload),
            Cmd::Nop => {} // Keepalive
            Cmd::Upd => {} // TODO: Flow control
        }

        Ok(())
    }

    fn handle_syn(&self, id: u32) -> Result<(), MuxError> {
        if self.streams.read().unwrap().contains_key(&id) {
            return Ok(()); // Duplicate SYN
        }

        let output = self.output.clone();
        let stream = Arc::new(Stream::new(
            id,
            Box::new(move |data| {
                // Use encode_with_payload to avoid intermediate allocations
                let encoded = Frame::encode_with_payload(Cmd::Psh, id, data);
                if let Err(e) = output(&encoded) {
                    eprintln!("Mux output error: {}", e);
                }
            }),
        ));

        self.streams.write().unwrap().insert(id, stream.clone());
        self.accept_queue.lock().unwrap().push(stream);

        Ok(())
    }

    fn handle_fin(&self, id: u32) {
        if let Some(stream) = self.streams.read().unwrap().get(&id) {
            stream.handle_fin();
        }
    }

    fn handle_psh(&self, id: u32, payload: &[u8]) {
        if let Some(stream) = self.streams.read().unwrap().get(&id) {
            stream.kcp_input(payload);
            stream.kcp_recv();
        }
    }

    fn send_syn(&self, id: u32) -> Result<(), MuxError> {
        self.send_frame(Frame::new(Cmd::Syn, id, vec![]))
    }

    pub fn send_fin(&self, id: u32) -> Result<(), MuxError> {
        self.send_frame(Frame::new(Cmd::Fin, id, vec![]))
    }

    fn send_frame(&self, frame: Frame) -> Result<(), MuxError> {
        if self.is_closed() {
            return Err(MuxError::MuxClosed);
        }

        (self.output)(&frame.encode()).map_err(|_| MuxError::OutputFailed)?;
        Ok(())
    }

    /// Remove a stream.
    pub fn remove_stream(&self, id: u32) {
        self.streams.write().unwrap().remove(&id);
    }

    /// Update all streams.
    pub fn update(&self, current: u32) {
        for stream in self.streams.read().unwrap().values() {
            stream.kcp_update(current);
            stream.kcp_recv();
        }
    }
}

/// Mux errors
#[derive(Debug)]
pub enum MuxError {
    MuxClosed,
    NoStreamAvailable,
    InvalidFrame,
    OutputFailed,
}

impl std::fmt::Display for MuxError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MuxError::MuxClosed => write!(f, "mux closed"),
            MuxError::NoStreamAvailable => write!(f, "no stream available"),
            MuxError::InvalidFrame => write!(f, "invalid frame"),
            MuxError::OutputFailed => write!(f, "output failed"),
        }
    }
}

impl std::error::Error for MuxError {}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicUsize, Ordering};

    #[test]
    fn test_mux_init() {
        let mux = Mux::new(
            MuxConfig::default(),
            true,
            Box::new(|_| Ok(())),
        );

        assert!(!mux.is_closed());
        assert_eq!(mux.num_streams(), 0);
    }

    #[test]
    fn test_mux_open_stream() {
        let output_count = Arc::new(AtomicUsize::new(0));
        let output_count_clone = output_count.clone();

        let mux = Mux::new(
            MuxConfig::default(),
            true,
            Box::new(move |_| {
                output_count_clone.fetch_add(1, Ordering::SeqCst);
                Ok(())
            }),
        );

        let stream = mux.open_stream().unwrap();
        assert_eq!(stream.id(), 1); // Client uses odd IDs
        assert_eq!(mux.num_streams(), 1);
        assert!(output_count.load(Ordering::SeqCst) > 0); // SYN was sent
    }

    #[test]
    fn test_stream_state() {
        let mux = Mux::new(
            MuxConfig::default(),
            true,
            Box::new(|_| Ok(())),
        );

        let stream = mux.open_stream().unwrap();
        assert_eq!(stream.state(), StreamState::Open);

        stream.close();
        assert_eq!(stream.state(), StreamState::LocalClose);
    }
}
